server/src/api/integration/integration.controller.ts

```diff
@@ -78,6 +78,56 @@ export const postConfigurations: RouteHandlerFnc = async (ctx) => {
   });
 };
 
+const withTokenRefresh = async (
+  ctx: IKoaContext,
+  resourceCall: () => Promise<any>,
+) => {
+  try {
+    return await resourceCall();
+  } catch (err) {
+    const provider = await oAuthProvider(ctx);
+    if (err instanceof InvalidTokenError && provider.tokenRefresh) {
+      const { roadmapId, integrationName } = ctx.params;
+      const userId = ctx.state.user.id;
+      const config = await integrationConfig(
+        integrationName,
+        Number(roadmapId),
+      );
+      const tokens = await integrationTokens(config, userId);
+      const { accessToken, refreshToken } = await provider.tokenRefresh(
+        tokens.refreshToken,
+        roadmapId,
+      );
+      await Token.transaction(async (trx) => {
+        // TODO: interface, class or something else to represent the token contents.
+        const upsertToken = async (type: string, value: string) => {
+          // TODO: Extract providers and token types somewhere to avoid magic numbers.
+          const newToken = {
+            forIntegration: config.id,
+            provider: integrationName,
+            instance: provider.instance,
+            user: userId,
+            type,
+            value,
+          };
+          const existing = await Token.query(trx).findOne({
+            forIntegration: newToken.forIntegration,
+            type: newToken.type,
+          });
+          if (existing) {
+            await existing.$query(trx).patchAndFetch(newToken);
+          } else {
+            await Token.query(trx).insert(newToken);
+          }
+        };
+        await upsertToken('access_token', accessToken);
+        await upsertToken('refresh_token', refreshToken);
+      });
+      return await resourceCall();
+    } else throw err;
+  }
+};
+
```

Comment:
When the token expires, this approach ends with one call returning InvalidTokenError (token expired) and the other two returning InvalidGrantError (the token is revoked as a new one has been fetched with refresh).  After a single set of fails, the token does get successfully refreshed and subsequent calls work, but it is a bit clunky right now.

Any ideas @juhoeerola, a refresh button on the config page or something else?

Reply:
I think we could sequence the api calls requiring oauth, such that one call must succeed before the others are even tried. This should handle the normal use cases.

There will still be cases where the token could expire while the user is on the page, or when multiple users trigger the refresh at the same time. However, I think regular page refresh (instead of explicit "refresh token" button) could be enough for the cases when the automatic refresh fails. (A token error could show error message suggesting a page refresh.)